File: src/java/src/com/luxio/q0001_TwoSum/TwoSum.java
package com.luxio.q0001_TwoSum;

import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

/**
 * 1.两数之和
 * 给定一个整数数组 nums和一个整数目标值 target，请你在该数组中找出 和为目标值 target 的那两个整数，并返回它们的数组下标。
 *
 * 你可以假设每种输入只会对应一个答案。但是，数组中同一个元素在答案里不能重复出现。
 *
 */
public class TwoSum {
    public static void main(String[] args) {
        int nums[] = {1, 3, 5, 7, 9, 4};
        int target = 9;
        int[] res = twoSum(nums, target);
        System.out.println(Arrays.toString(res));
    }

    private static int[] twoSum(int [] nums, int target) {
        Map<Integer, Integer> map = new HashMap<>();
        for (int i = 0; i < nums.length; i++) {
            if (map.containsKey(target - nums[i])) {
                return new int[]{map.get(target - nums[i]), i};
            }
            map.put(nums[i], i);
        }
        return new int[0];
    }
}

File: src/0001.two_sum/two_sum.cpp
#include <vector>

#include "two_sum.h"
#include <unordered_map>

using namespace std;

vector<int> twoSum(vector<int>& nums, int target) 
{
    unordered_map<int, int> map;
    int n = nums.size();
    for (int i = 0; i < n; i++)
    {
        auto it = map.find(target - nums[i]);
        if (it != map.end()) {
            return { it->second, i };
        }
        map[nums[i]] = i;
    }
    return {};
}

/*
vector<int> twoSum(vector<int>& nums, int target) {
    int n = nums.size();
    for (int i = 0; i < n; ++i) {
        for (int j = i + 1; j < n; ++j) {
            if (nums[i] + nums[j] == target) {
                return { i, j };
            }
        }
    }
    return {};
}
*/
